File: datapipeline/data_report.py


# def get_data_report_data():
#     connection = get_db()
#     output = {}
#     sql = ''' 
# select 
#     n_companies, 
#     n_companies_with_orders 

# from companies_with_orders 

# order by timestamp desc 

# limit 1 '''
#     df = query_database(connection, sql)
#     web_dict = to_web_dict(df)
#     web_dict['data'] = web_dict['data'][0]
#     output['companyOrderSummary'] =  web_dict

#     sql = ''' 
# select 
#     n_companies, 
#     n_companies_with_export_countries

# from companies_with_export_countries

# order by timestamp desc 

# limit 1 '''
#     df = query_database(connection, sql)
#     web_dict = to_web_dict(df)
#     web_dict['data'] = web_dict['data'][0]
#     output['companyExportCountriesSummary'] = web_dict

#     sql = ''' 
# select 
#     n_companies, 
#     n_companies_with_countries_of_interest

# from companies_with_countries_of_interest

# order by timestamp desc 

# limit 1 '''
#     df = query_database(connection, sql)
#     web_dict = to_web_dict(df)
#     web_dict['data'] = web_dict['data'][0]
#     output['countriesOfInterestSummary'] = web_dict

#     order_frequencies = {}
#     sql = ''' 
# with latest_timestamp as (
#   select max(timestamp) as timestamp from order_frequency_by_date order by timestamp desc limit 1
# ), order_frequency as (
#   select * from order_frequency_by_date join latest_timestamp using (timestamp)
# )
# select 
#     date, 
#     count as n_orders

# from order_frequency

# order by date

# '''
#     df = query_database(connection, sql)
#     web_dict = to_web_dict(df)
#     order_frequencies['daily'] = web_dict

#     sql = ''' 
# with latest_timestamp as (
#   select max(timestamp) as timestamp from order_frequency_by_date order by timestamp desc limit 1
# ), order_frequency as (
#   select * from order_frequency_by_date join latest_timestamp using (timestamp)
# )
# select 
#     date_trunc('week', date) as date, 
#     sum(count) as n_orders

# from order_frequency

# group by 1

# order by date

# '''
#     df = query_database(connection, sql)
#     web_dict = to_web_dict(df)
#     order_frequencies['weekly'] = web_dict

#     sql = ''' 
# with latest_timestamp as (
#   select max(timestamp) as timestamp from order_frequency_by_date order by timestamp desc limit 1
# ), order_frequency as (
#   select * from order_frequency_by_date join latest_timestamp using (timestamp)
# )
# select 
#     date_trunc('month', date) as date, 
#     sum(count) as n_orders

# from order_frequency

# group by 1

# order by date

# '''
#     df = query_database(connection, sql)
#     web_dict = to_web_dict(df)
#     order_frequencies['monthly'] = web_dict

#     output['orderFrequency'] = order_frequencies
    
#     return output

# @app.route('/api/get-matched-companies')
# @hawk_required
# def get_matched_companies():
#     sql = ''' select * from matched_companies order by timestamp limit 1 '''
#     db = get_db()
#     rows = query_db(db, sql)
#     row = rows[0]
#     return {
#         'nCompanies': row[0],
#         'nMatches': row[1],
#         'nDuplicates': row[2],
#         'nUniqueMatches': row[3]
#     }

# @app.route('/api/get-sector-matches')
# @hawk_required
# def get_sector_matches():
#     sql = ''' select * from sector_matches order by timestamp limit 1 '''
#     db = get_db()
#     rows = query_db(db, sql)
#     row = rows[0]
#     return {
#         'nCompanies': row[0],
#         'nSectors': row[1],
#         'nMatches': row[2],
#     }

# @app.route('/api/get-top-sectors')
# @hawk_required
# def get_top_sectors():
#     sql = ''' with latest_top_sectors as (
#       select max(timestamp) as timestamp_most_recent from top_sectors

#     ), top_sectors_most_recent as (
#       select
#         sector, 
#         count

#       from top_sectors join latest_top_sectors on 1=1 

#       where timestamp = timestamp_most_recent
#     )
#     select * from top_sectors_most_recent order by count desc '''
#     db = get_db()
#     rows = query_db(db, sql)
#     return {
#         'data': [{'name': row[0], 'count': row[1]} for row in rows]
#     }

File: app/static/data_report.js
import React from 'react';
import axios from 'axios';
import * as d3 from "d3";
import { render } from 'react-dom';

window.React = React;

class App extends React.Component {

    constructor(props) {
        super(props);
        this.orderFrequencyChartId = 'order-frequency-chart';
        this.state = {
            data: null,
            nRowsTopSector: 5
        };
        this.drawOrderFrequencyChart = this.drawOrderFrequencyChart.bind(this);
        this.onNRowsTopSectorChange = this.onNRowsTopSectorChange.bind(this);
    }

    componentDidMount() {
        axios.get('/api/v1/get-data-report-data')
            .then(response => response.data)
            .then(data => this.setState({data: data}))
            .then(x => this.drawOrderFrequencyChart());
    }

    drawOrderFrequencyChart() {
        if (this.state.data == null){
            return
        }
        const data = this.state.data.omisOrderFrequency.values;
        const dates = data.map(row=>new Date(row[0]));
        const n_orders = data.map(row=>row[1]);
        const parsedData = d3.zip(dates, n_orders);

        let container = d3.select(`#${this.orderFrequencyChartId}`);
        let canvas = container.select(".canvas");
        let canvasWidth = parseInt(container.style('width'), 10);
        let canvasHeight = 300;
        canvas.attr("width", canvasWidth)
            .attr("height", canvasHeight);
        
        let chart = canvas.select(".chart");
        let chartMargins = {
            left: 50,
            right: 30,
            bottom: 50,
            top: 15
        };
        let chartWidth = canvasWidth - chartMargins.left - chartMargins.right;
        let chartHeight = canvasHeight - chartMargins.top - chartMargins.bottom;
        chart.attr(
            'transform',
            `translate(${chartMargins.left}, ${chartMargins.top})`
        );

        
        let xScale = d3.scaleTime()
            .range([0, chartWidth])
            .domain([d3.min(dates), d3.max(dates)]);

        let yScale = d3.scaleLinear()
            .range([chartHeight, 0])
            .domain([d3.min(n_orders), d3.max(n_orders)]);

        chart.selectAll("*").remove();

        var valueline = d3.line()
            .x(function(d) { return xScale(d[0]); })
            .y(function(d) { return yScale(d[1]);  })
            .curve(d3.curveMonotoneX);

        chart.append("path")
            .data([parsedData]) 
            .attr("class", "line")  
            .attr("d", valueline);
        
        let xAxis = d3.axisBottom(xScale);
        chart.append('g')
            .attr('transform', `translate(0, ${chartHeight})`)
            .call(xAxis);

        let yAxis = d3.axisLeft(yScale);
        chart.append('g')
            .call(yAxis);

        chart.selectAll(".dot")
            .data(parsedData)
            .enter()
            .append("circle") 
            .attr("class", "dot") 
            .attr("cx", function(d) { return xScale(d[0]); })
            .attr("cy", function(d) { return yScale(d[1]); })
            .attr("r", 2);  

    }
    
    onNRowsTopSectorChange(nRows){
        this.setState({nRowsTopSector: nRows});
    }
    
    render() {
        console.log("render");
        console.log(this.state);

        const {data, nRowsTopSector, orderFrequency} = this.state;
        if (data == null){
            return '';
        }
        
        const {
            nCompanies,
            nCompaniesMatchedToCompaniesHouse,
            nCompaniesMatchedToSector,
            nCompaniesMatchedToDuplicateCompaniesHouse,
            nSectors,
            nCompaniesWithOmisOrders,
            nCompaniesWithExportCountries,
            nCompaniesWithFutureInterestCountries,
        } = data;
        
        const chartId = this.orderFrequencyChartId;
        return (
            <div style={{paddingTop: '1em'}}>
              <MatchedCompanies
                nCompanies={nCompanies}
                nDuplicates={nCompaniesMatchedToDuplicateCompaniesHouse}
                nMatches={nCompaniesMatchedToCompaniesHouse}
              />
              <SectorMatches
                nCompanies={nCompanies}
                nMatches={nCompaniesMatchedToSector}
                nSectors={nSectors}
              />
              <TopSectors
                data={data.topSectors.values}
                onChange={this.onNRowsTopSectorChange}
                nRows={nRowsTopSector}
              />
              <CompanyOrderSummary nCompanies={nCompanies} nWithOrders={nCompaniesWithOmisOrders}/>
              <CompanyExportCountriesSummary
                nCompanies={nCompanies}
                nWithExportCountries={nCompaniesWithExportCountries}
              />
              <CompanyCountriesOfInterestSummary
                nCompanies={nCompanies}
                nWithCountriesOfInterest={nCompaniesWithFutureInterestCountries}
              />
              <OrderFrequency
                chartId={chartId}
                data={data.omisOrderFrequency.values}
              />
            </div>
        );

    }
}

const MatchedCompanies = ({nCompanies, nDuplicates, nMatches}) => {
    if(nCompanies == null){
        return '';
    }

    let nUniqueMatches = nMatches - nDuplicates;
    
    return (
        
        <div className="data-report">
          <h3 style={{paddingBottom: '0.5em'}}>Datahub - Companies House matching</h3>
          <table className="table table-striped">
            <thead>
              <tr>
                <th></th>
                <th>Count</th>
                <th>Percent</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>Companies</td>
                <td>{nCompanies}</td>
                <td></td>
              </tr>
              <tr>
                <td>Matches</td>
                <td>{nMatches}</td>
                <td>{Math.round(100 * 100 * nMatches/nCompanies)/100}</td>
              </tr>
            </tbody>
          </table>

          <h3 style={{paddingBottom: '0.5em'}}>Breakdown of matched companies</h3>
          <table className="table table-striped">
            <thead>
              <tr>
                <th></th>
                <th>Count</th>
                <th>Percent</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>Matches</td>
                <td>{nMatches}</td>
                <td></td>
              </tr>
              <tr>
                <td>Unique Matches</td>
                <td>{nUniqueMatches}</td>
                <td>{Math.round(100 * 100 * nUniqueMatches / nMatches) / 100}</td>
              </tr>
              <tr>
                <td>Duplicates</td>
                <td>{nDuplicates}</td>
                <td>{Math.round(100 * 100 * nDuplicates / nMatches) / 100}</td>
              </tr>
            </tbody>
          </table>
        </div>
        
    );
};

const SectorMatches = ({nCompanies, nSectors, nMatches}) => {

    if(nCompanies == null){
        return '';
    }
    
    return (
        <div className="sector-summary">
          <h3>Sector summary</h3>
          <table className="table table-striped">
            <tbody>
              <tr>
                <td>#Companies</td><td>{nCompanies}</td>
              </tr>
              <tr>
                <td>#Sectors</td><td>{nSectors}</td>
              </tr>
              <tr>
                <td>#Companies matched to a sector</td><td>{nMatches}</td>
              </tr>
              <tr>
                <td>%Companies matched to a sector</td>
                <td>{Math.round(100 * 100 * nMatches/nCompanies)/100}</td>
              </tr>
            </tbody>
          </table>
        </div>
    );
};

const TopSectors = ({data, onChange, nRows=5}) => {
    let nRowsInput;
    const onChange2 = (e) => {
        return onChange(nRowsInput.value);
    };
    const filteredData = data.slice(0, nRows);
    return (
        <div>
          <h3>Top sectors</h3>
          <div className="form-group">
            <label className="form-label">Number of rows</label>
            <input
              ref={input => nRowsInput = input}
              onChange={onChange2}
              className="form-control"
              type="number"
              value={nRows}
            />
          </div>
          <table className="table table-striped">
            <thead>
              <tr>
                <th>Sector</th>
                <th>Count</th>
              </tr>
            </thead>
            <tbody>
              {
                  filteredData.map((sector, i) => {
                      return (
                          <tr key={i}>
                            <td>{sector[0]}</td>
                            <td>{sector[1]}</td>
                          </tr>
                      );
                  })
              }
            </tbody>
          </table>
        </div>        
    );
};

const CompanyOrderSummary = ({nCompanies, nWithOrders}) => {
    return (
        <div className="company-order-summary">
          <h3>Company order summary</h3>
          <table className="table table-striped">
            <tbody>
              <tr><td>#Companies</td><td>{nCompanies}</td></tr>
              <tr><td>#Companies with orders</td><td>{nWithOrders}</td></tr>
              <tr>
                <td>%Companies with orders</td>
                <td>{Math.round(100 * 100 * nWithOrders/nCompanies)/100}</td>
              </tr>
            </tbody>
          </table>
        </div>
    );

};

const CompanyExportCountriesSummary = ({nCompanies, nWithExportCountries}) => {
    if(nCompanies == null){
        return '';
    }
    
    return (
        <div className="company-export-country-summary">
          <h3>Company export country summary</h3>
          <table className="table table-striped">
            <tbody>
              <tr><td>#Companies</td><td>{nCompanies}</td></tr>
              <tr><td>#Companies with export countries</td><td>{nWithExportCountries}</td></tr>
              <tr>
                <td>%Companies with export countries</td>
                <td>{Math.round(100 * 100 * nWithExportCountries/nCompanies)/100}</td>
              </tr>
            </tbody>
          </table>
        </div>
    );

};

const CompanyCountriesOfInterestSummary = ({nCompanies, nWithCountriesOfInterest}) => {
    if(nCompanies == null){
        return '';
    }

    return (
        <div className="company-countries-of-interest-summary">
          <h3>Countries of interest summary</h3>
          <table className="table table-striped">
            <tbody>
              <tr><td>#Companies</td><td>{nCompanies}</td></tr>
              <tr>
                <td>#Companies with countries of interest</td>
                <td>{nWithCountriesOfInterest}</td>
              </tr>
              <tr>
                <td>%Companies with countries of interest</td>
                <td>{Math.round(100 * 100 * nWithCountriesOfInterest/nCompanies)/100}</td>
              </tr>
            </tbody>
          </table>
        </div>
    );
};

const OrderFrequency = ({chartId, data}) => {
    if(data == null){
        return '';
    }

    return (
        <div className='order-frequency'>
          <h3>Order frequency</h3>
          <OrderFrequencyChart data={data} id={chartId}/>
        </div>
    );
};

const OrderFrequencyChart = ({data, id}) => {
    if(data == null){
        return '';
    }

    return (
        <div id={id} className="order-frequency-chart">
          <svg className="canvas">
            <g className="chart">
            </g>
          </svg>
        </div>
    );
};

render(<App/>, document.getElementById('react-container'));
